tests/src/runtimeApi/module/hipLaunchCooperativeKernel.cpp

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright (c) 2019 - present Advanced Micro Devices, Inc. All rights reserved.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+// Simple test for hipLaunchCooperativeKernel API.
+
+/* HIT_START
+ * BUILD: %t %s ../../test_common.cpp EXCLUDE_HIP_PLATFORM nvcc
+ * TEST: %t
+ * HIT_END
+ */
+
+#include "hip/hip_runtime.h"
+#include "hip/hip_runtime_api.h"
+#include "hip/hcc_detail/device_library_decls.h"
+#include <iostream>
+#include <chrono>
+#include "test_common.h"
+
+using namespace std::chrono;
+
+const static uint BufferSizeInDwords = 448 * 1024 * 1024;
+
+__global__ void test_gws(uint* buf, uint bufSize, long* tmpBuf, long* result)
+{
+    extern __shared__ long tmp[];
+    uint offset = blockIdx.x * blockDim.x + threadIdx.x;
+    uint stride = gridDim.x  * blockDim.x;
+    uint nwm1   = gridDim.x  * gridDim.y * gridDim.z - 1;
+
+    long sum = 0;
+    for (uint i = offset; i < bufSize; i += stride) {
+        sum += buf[i];
+    }
+    tmp[threadIdx.x] = sum;
+
+    __syncthreads();
+
+    if (threadIdx.x == 0) {
+        sum = 0;
+        for (uint i = 0; i < blockDim.x; i++) {
+            sum += tmp[i];
+        }
+        tmpBuf[blockIdx.x] = sum;
+    }
+
+    // TODO replace this line with the sync function once it's available
+    __ockl_gws_barrier(nwm1, 0);
```

Comment:
Note that this function does not guarantee that all the updates to tmpBuf will be visible when the barrier is passed.  If this produces the expected sum, it is just luck.